x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/WildcardServiceProviderResolver.java

```diff
@@ -0,0 +1,232 @@
+
+        if (Files.exists(path)) {
+            try {
+                resolver.reload(path);
+            } catch (IOException e) {
+                throw new ElasticsearchException("File [{}] (from setting [{}]) cannot be loaded",
+                    e, path.toAbsolutePath(), FILE_PATH_SETTING.getKey());
+            }
+        } else if (FILE_PATH_SETTING.exists(environment.settings())) {
+            // A file was explicitly configured, but doesn't exist. That's a mistake...
+            throw new ElasticsearchException("File [{}] (from setting [{}]) does not exist",
+                path.toAbsolutePath(), FILE_PATH_SETTING.getKey());
+        }
+
+        final FileWatcher fileWatcher = new FileWatcher(path);
+        fileWatcher.addListener(new FileChangesListener() {
+            @Override
+            public void onFileCreated(Path file) {
+                onFileChanged(file);
+            }
+
+            @Override
+            public void onFileDeleted(Path file) {
+                onFileChanged(file);
+            }
+
+            @Override
+            public void onFileChanged(Path file) {
+                try {
+                    resolver.reload(file);
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
+        });
+        try {
+            resourceWatcherService.add(fileWatcher);
+        } catch (IOException e) {
+            throw new ElasticsearchException("Failed to watch file [{}] (from setting [{}])",
+                e, path.toAbsolutePath(), FILE_PATH_SETTING.getKey());
+        }
+        return resolver;
+    }
+
+    public SamlServiceProvider resolve(String entityId, String acs) {
+        final State currentState = stateRef.get();
+
+        Tuple<String, String> cacheKey = new Tuple<>(entityId, acs);
+        final SamlServiceProvider cached = currentState.cache.get(cacheKey);
+        if (cached != null) {
+            logger.trace("Service for [{}] [{}] is cached [{}]", entityId, acs, cached);
+            return cached;
+        }
+
+        final Map<String, SamlServiceProvider> matches = new HashMap<>();
+        currentState.services.forEach((name, wildcard) -> {
+            final SamlServiceProviderDocument doc = wildcard.apply(scriptService, entityId, acs);
+            if (doc != null) {
+                final SamlServiceProvider sp = serviceProviderFactory.buildServiceProvider(doc);
+                matches.put(name, sp);
+            }
+        });
+
+        switch (matches.size()) {
+            case 0:
+                logger.trace("No wildcard services found for [{}] [{}]", entityId, acs);
+                return null;
+
+            case 1:
+                final SamlServiceProvider serviceProvider = Iterables.get(matches.values(), 0);
+                logger.debug("Found exactly 1 wildcard service for [{}] [{}] - [{}]", entityId, acs, serviceProvider);
```

Comment:
Maybe make this `TRACE` since we log pretty much the same information in `DEBUG` in `private void resolveWildcardService(String entityId, String acs, ActionListener<SamlServiceProvider> listener) `